server/controllers/postgres.js

```diff
@@ -1,18 +1,19 @@
-// TODO add table to track latest id to automate inserting consistently retrievable ids
-// TODO add table to track total count of all records
 // TODO validate and sanitize data
 
 const { getProduct, updateProduct, getKeys, deleteProduct, insertProduct, getLatest } = require('../../database/postgres/database.js');
 const { asyncHandler, ErrAPI, generateIds } = require('./utils.js');
 const { randomDoc } = require('../../database/postgres/data.js');
 const md5 = require('md5');
+const { setCache, invalidateKey } = require("./redis.js");
 
 // keys for info values to control entries
 const infoKeys = ['brand', 'itemColor', 'itemName', 'isPrimeFreeOneDay', 'isFreeDelivery'];
 
 // get a single or batch of product descriptions
 exports.getProduct = asyncHandler( async(req, res, next) => {
+  const { productId } = req.params;
   const product = await getProduct(req.hash);
+  await setCache(productId, JSON.stringify(product));
```

Comment:
so every time a productId is retrieved from the database, you store the results from the database in the redis cache for fast access if this same productId is needed again in the future, then you return the product information to the client. Nice. I like how modular all of this code is.